fix: add custom CA certificate support for login flows

3axap4eHko opened the pull request:
## What

  Adds support for custom CA certificates in OAuth login flows to fix authentication failures behind corporate
  proxies with SSL/TLS inspection.

  ## Why

  The Codex CLI fails to authenticate when running behind corporate proxies that perform SSL/TLS inspection with
  custom CA certificates. The OAuth token exchange fails because `reqwest::Client::new()` only trusts system
  default certificate roots, causing it to reject connections intercepted by corporate proxies presenting
  certificates signed by custom CAs.

  This affects all Codex CLI users in enterprise environments with:
  - SSL/TLS inspection proxies (Zscaler, Palo Alto Networks, etc.)
  - Custom internal CA certificates
  - Air-gapped environments with internal certificate authorities

  Fixes #6849

  ## How

  **Implementation:**
  - Added `build_login_http_client()` function that creates an HTTP client with custom CA certificate support
  - Reads CA certificate path from environment variables (priority order):
    1. `CODEX_CA_CERTIFICATE` (primary)
    2. `SSL_CERT_FILE` (fallback, standard across many tools)
  - Updated all three authentication flows to use the custom client:
    - `exchange_code_for_tokens()` (OAuth code exchange)
    - `obtain_api_key()` (API key exchange)
    - `run_device_code_login()` (device code flow)
  - Follows the same pattern as existing `otel_provider.rs` implementation

  **Testing:**
  - Added comprehensive test coverage with 3 new tests:
    - Certificate loading via `CODEX_CA_CERTIFICATE` env var
    - Certificate loading via `SSL_CERT_FILE` fallback
    - Invalid certificate rejection with proper error handling
  - All existing tests continue to pass
  - Added `serial_test` dependency to safely test env var manipulation

  **Usage:**
  ```bash
  export CODEX_CA_CERTIFICATE=/path/to/corporate-ca.pem
  codex login
  ```
  or using the standard env var:
  ```bash
  export SSL_CERT_FILE=/path/to/corporate-ca.pem
  codex login
  ````
  
  **Testing:**

  - ✅ All tests pass (cargo test)
  - ✅ No clippy warnings (cargo clippy --tests -- -D warnings)
  - ✅ Code formatted (cargo fmt --check)
  - ✅ Tested with both environment variables
  - ✅ Invalid certificate properly rejected with clear error message

3axap4eHko committed:
Add custom CA support for login flows

etraut-openai commented:
@codex review

3axap4eHko committed:
fix: adds support of multiple certificates per pem file

etraut-openai commented:
@3axap4eHko, thanks for the contribution! It looks like there are some minor code format issues. Please fix those when you have time.

joshka-oai reviewed (commented):
Brief review notes / questions to help work out whether to do a deeper review.

bolinfest commented on codex-rs/login/src/server.rs:
```diff
@@ -719,3 +799,214 @@ pub(crate) async fn obtain_api_key(
+
+    fn restore_env(key: &str, value: Option<String>) {
+        match value {
+            Some(val) => unsafe { env::set_var(key, val) },
```
`set_var` is indeed `unsafe` because `cargo test` can run multiple tests concurrently in the same process, so multiple threads calling `set_var` at the same time can result in undefined behavior.

Generally, I think there are two options:

- refactor code so the logic takes the "environment" as an abstraction
- create a helper binary that exercises the library code you want to test and spawn it from this test with the desired env

joshka-oai commented:
I'm assuming that your environment sets some sort of CA root authority that your general tools (browser etc.) properly check, so you don't tend to get errors in non-cli tooling. Is this something that we could generally load from the environment in the same way without having to manually configure this with environment variables? I.e. perhaps configuring the tooling to load from the OS Cert store where this is setup. There's probably a crate that sets this up automatically if that's the case. I'm not sure whether this is relevant to the use cases here, so please forgive the naive question.

joshka-oai commented on codex-rs/login/src/server.rs:
```diff
@@ -719,3 +799,214 @@ pub(crate) async fn obtain_api_key(
+        unsafe {
+            env::set_var(CODEX_CA_CERT_ENV, &cert_path);
+            env::remove_var(SSL_CERT_FILE_ENV);
+        }
```
@codex go deeper and find any code that would be affected by this problem

3axap4eHko committed:
docs: add CA certificate environment variable documentation

joshka-oai commented on codex-rs/login/src/server.rs:
```diff
@@ -719,3 +799,214 @@ pub(crate) async fn obtain_api_key(
+    use std::env;
+    use tempfile::TempDir;
+
+    const TEST_CERT_1: &str = "-----BEGIN CERTIFICATE-----
```
Sounds good - thanks for the push back. I agree with your assessment.

etraut-openai commented:
@3axap4eHko, I wanted to give you an update on this PR. Since it involves security-sensitive code, we have pulled in some security experts outside of the codex team to provide a more thorough code review. This is a holiday week in the U.S. (Thanksgiving), so many OpenAI employees are out of the office. That means we probably won't make any further progress on this until next week.

joshka-oai commented:
After chatting about this with our internal security, we'd prefer to have the settings for this added to the config.toml file rather than as an environment variable. If you're concerned that you need single call setups (where you don't want to set this at the entire machine level) codex has a `--config` flag that can generally be used to pass config key/values like this. We don't need to add a specific command line flag for this. Let's call this setting `login_ca_certificate`.

Let's use rustls-pki-types to open / parse the pem files rather thank hand coded.

Happy to continue the implementation on this with codex, or let you complete it. The things I'd generally want to make sure are covered here:
- the rationale that covers why this setting is necessary is covered in the code as a doc comment on the code, and in the configuration docs (https://github.com/openai/codex/blob/main/docs/config.md#authentication-and-authorization)
- tests (the ones you have look good - modify to be config based instead of environment variable based)

joshka-oai committed:
login: harden custom CA handling and isolate env-based tests

Enterprise TLS inspection proxies use custom roots, so OAuth token exchanges
fail when we only trust system CAs. This change switches PEM parsing to
rustls-pki-types (multi-cert bundles included) and surfaces clearer, user-
facing errors that explain how to fix invalid or empty CA files via
CODEX_CA_CERTIFICATE/SSL_CERT_FILE.

To avoid cross-test races with process-wide env vars, CA path selection now
uses a small EnvSource abstraction in unit tests, and environment-dependent
behavior is verified via an assert_cmd-driven login_ca_probe helper binary.
This keeps normal tests isolated while still validating env precedence and
error messaging.

Also updates login dev-deps (assert_cmd/pretty_assertions), removes serial_test,
and re-exports build_login_http_client for the probe helper.

3axap4eHko commented:
@joshka-oai Oh thanks for helping, I didn't expect that. I was a bit off for holidays

3axap4eHko committed:
login: support trusted cert bundles and make CA tests env-safe

3axap4eHko committed:
login: tolerate CRL blocks in CA bundles and test

3axap4eHko committed:
tests: fix login fixtures

3axap4eHko commented:
@codex review

joshka-oai commented on codex-rs/login/src/server.rs:
```diff
@@ -491,6 +499,148 @@ pub(crate) struct ExchangedTokens {
+    if let Some(path) = login_ca_certificate_path(env_source) {
+        let certificates = read_ca_certificates(&path)?;
+
+        for (idx, cert) in certificates.iter().enumerate() {
```
You can safely ignore this one. The AGENTS.md advice needs to be changed on this.

3axap4eHko commented:
@joshka-oai is there something here that needs to be addressed?

joshka-oai commented:
Hey there, I haven't forgotten about this. Thanks for hanging in there and putting up with the delay on this. It's just below my threshold for where I can allocate time to get this over the line, and the majority of the team has been taking a break over the holiday period. I'll talk with a few people early next week to get some extra eyes on this.

Something to note is the changes coming in reqwest 0.13, which moves to using rustls by default and has some extra things around merging tls certs. I haven't looked into that change yet to see if there's overlap on this, but if there is, then we likely should upgrade reqwest first and then rebase this on top of those newer approaches.
 https://seanmonstar.com/blog/reqwest-v013-rustls-default/

3axap4eHko commented:
> Hey there, I haven't forgotten about this. Thanks for hanging in there and putting up with the delay on this. It's just below my threshold for where I can allocate time to get this over the line, and the majority of the team has been taking a break over the holiday period. I'll talk with a few people early next week to get some extra eyes on this.
> 
> Something to note is the changes coming in reqwest 0.13, which moves to using rustls by default and has some extra things around merging tls certs. I haven't looked into that change yet to see if there's overlap on this, but if there is, then we likely should upgrade reqwest first and then rebase this on top of those newer approaches. https://seanmonstar.com/blog/reqwest-v013-rustls-default/

I'll review changes in reqwest

joshka-oai commented:
I took a bit of a look at reqwest 0.13. We can update to 0.13 when https://github.com/open-telemetry/opentelemetry-rust/pull/3293 is released.

3axap4eHko committed:
fix(login): reconcile rebase fallout
